### DISCUSSION TITLE
Custom Spring Boot HealthCheck to report the status of tasks not working as expected

### DISCUSSION THREAD

**Discussion by Yith1** [kagkarlsson/db-scheduler#563] [General]
I've written a Spring Boot Healthcheck to return the status of the tasks I have running. They are recurring tasks with a frequency of 10 minutes.
![image](https://github.com/user-attachments/assets/bde42406-355a-4cf6-b984-93cc930578b1)
![image](https://github.com/user-attachments/assets/4edcd105-6fe6-4722-8bc4-2c7cd6e913e2)

The idea is for it to fail when there are failures occurring in the tasks.

It's not quite working as expected though. Once a failure occurs the healthcheck fails forever.
This is because theDuration.ofMinutes(1)); call at line 44 is not doing what I would expect.
There is no documentation so I assumed that the duration should be a small one. However I see that in the sql that is used for the method call that is actually used to check whether the last success was greater than that interval in the past, which it almost always will be in my case since the frequency is 10 minutes.

![image](https://github.com/user-attachments/assets/87a6a0c9-6c1f-480b-a660-88dfd16763a2)

So I could set the interval to greater than 10 minutes, but then that would not do what I want since failures would not be reported by the healthcheck until after 10 minutes has passed.

Instead I think I need this change to my code:
![image](https://github.com/user-attachments/assets/e1acba20-162e-4b95-8b91-090684d5dcd2)
and also: 
![image](https://github.com/user-attachments/assets/be0dc5ea-975d-4987-8fa7-4653a7f6e93d)

By checking that the most recent success is more recent than the most recent failure then I can be sure of the status of my tasks and with keeping the interval as it is at 1 minute (or maybe less) then I can be sure that I report my failure state quickly and accurately.

My question is though, why does the scheduler.getFailingExecutions() use the sql it does. Why does it not compare the last_failure and last_success dates, or maybe even look for consecutive_failures > 0 ?

Or is there a better way to write a healthcheck like I would like?

**Comment by kagkarlsson** [comment]
I think this must be a bug in `getExecutionsFailingLongerThan(...)`. Strange it has survived for so long! 

Is this better?

```sql
where consecutive_failures > 0 and (last_success is null or last_success < ?)) "
```

I suspect this code might have been written before we started tracking consecutive_failures...

**Comment by Yith1** [comment]
I'm not really sure what the best fix is. It really depends on what exactly you intend the method to do.
